pkg/exchange/coinbase/parse.go

```diff
@@ -0,0 +1,101 @@
+package coinbase
+
+import (
+	"encoding/json"
+	"errors"
+)
+
+// See https://docs.cdp.coinbase.com/exchange/docs/websocket-channels for message types
+func (s *Stream) parseMessage(data []byte) (interface{}, error) {
+	var msgType string
+	{
+		var e messageBaseType
+		json.Unmarshal(data, &e)
+		msgType = e.Type
+	}
+
+	switch msgType {
+	case "heartbeat":
+		var msg HeartbeatMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
+	case "status":
+		var msg StatusMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
+	case "auction":
+		var msg AuctionMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
+	case "rfq_match":
+		var msg RfqMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
+	case "ticker":
+		var msg TickerMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
+	case "received":
+		// try market order first
+		{
+			var msg ReceivedMarketOrderMessage
+			json.Unmarshal(data, &msg)
+			if !msg.Funds.IsZero() {
+				return &msg, nil
+			}
+		}
+		var msg ReceivedLimitOrderMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
+	case "open":
+		var msg OpenMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
+	case "done":
+		var msg DoneMessage
+		json.Unmarshal(data, &msg)
+		return &msg, nil
```

Comment:
1. Each json.Unmarshal should check for errors

Reply:
Done.